docs: move deploy docs to docs/guides (#674)
ai-dynamo/dynamo: docs/guides/dynamo_cloud.md

<!--
SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

> 

Reviewer: All the other md files in guides are verbs associated with the CLI. Should we call this one `dynamo_deploy` too. I don't want people to think dynamo deploy and cloud are distinct.

Author: Hmm then people might get confused with the folder. I can put this file into the dynamo_deploy folder though!

Author: Moved it

Reviewer: nah, lets just keep as is then

> # Dynamo Cloud Deployment Platform

Reviewer: ```suggestion
# Dynamo Deploy: Dynamo Cloud Kubernetes Platform
```

> The Dynamo cloud platform is a comprehensive solution for deploying and managing Dynamo inference graphs in Kubernetes environments. It provides a streamlined experience for deploying, scaling, and monitoring your inference services.

Reviewer: ```suggestion
The Dynamo cloud platform is a comprehensive solution for deploying and managing Dynamo inference graphs (also referred to as pipelines) in Kubernetes environments. It provides a streamlined experience for deploying, scaling, and monitoring your inference services.
```

## Overview

The Dynamo cloud platform consists of several key components:

> - **Dynamo Operator**: A Kubernetes operator that manages the lifecycle of Dynamo inference graphs

Reviewer: ```suggestion
- **Dynamo Operator**: A Kubernetes operator that manages the lifecycle of Dynamo inference graphs from build -> deploy
```

> - **API Store**: Stores and manages service configurations and metadata

Reviewer: ```suggestion
- **API Store**: Stores and manages service configurations and metadata related to Dynamo deployments. Needs to be exposed externally.
```

- Access to a container registry (e.g., Docker Hub, NVIDIA NGC, etc.)
- `kubectl` configured to access your cluster
- Helm installed (version 3.0 or later)

## Building Docker Images for Dynamo Cloud Components

The Dynamo cloud platform components need to be built and pushed to a container registry before deployment. You can build these components individually or all at once.

### Setting Up Environment Variables

First, set the required environment variables for building and pushing images:

```bash
# Set your container registry and organization
export CI_REGISTRY_IMAGE=<CONTAINER_REGISTRY>/<ORGANIZATION>
# Set the image tag (e.g., latest, 0.0.1, etc.)
export CI_COMMIT_SHA=<TAG>
```

Where:
- `<CONTAINER_REGISTRY>/<ORGANIZATION>`: Your container registry and organization name
  - Examples: `nvcr.io/myorg`, `docker.io/myorg`
- `<TAG>`: The version tag for your images
  - Examples: `latest`, `0.0.1`, `v1.0.0`

> [!IMPORTANT]
> Make sure you're logged in to your container registry before pushing images:
> ```bash
> docker login <CONTAINER_REGISTRY>
> ```

### Building Components

You have two options for building the components:

#### Option 1: Build All Components at Once

> This is the simplest approach and builds all components in a single command:

Reviewer: ```suggestion
This is the simplest approach and builds and pushes all components in a single command:
```

### Prerequisites

Make sure you're in the correct directory:
```bash
cd deploy/dynamo/helm
```

Set your namespace (this will be used for all deployments):
```bash
export KUBE_NS=hello-world    # Change this to your preferred namespace
```

### Deployment Steps

1. **Create Namespace and Set Context**

```bash
# Create a new namespace
kubectl create namespace $KUBE_NS

# Set the namespace as your default context
kubectl config set-context --current --namespace=$KUBE_NS

# [Optional] Create image pull secrets if your registry requires authentication
kubectl create secret docker-registry docker-imagepullsecret \
  --docker-server=<your-registry> \
  --docker-username=<your-username> \
  --docker-password=<your-password> \
  --namespace=$KUBE_NS
```

2. **Deploy Using the Helm Chart**

Set the required environment variables:
```bash
export NGC_TOKEN=$NGC_API_TOKEN
export NAMESPACE=$KUBE_NS
export CI_COMMIT_SHA=<TAG>  # Use the same tag you used when building the images
export CI_REGISTRY_IMAGE=<CONTAINER_REGISTRY>/<ORGANIZATION>  # Use the same registry/org you used when building the images
export RELEASE_NAME=$KUBE_NS
```

Deploy the platform:
```bash
./deploy.sh
```

> 

Reviewer: Let's add some logic here on exposing the `api-store` externally. Something like this
```suggestion
3. Expose Dynamo Cloud externally

You must also expose the `dynamo-store` service within the namespace externally. This will be the endpoint the CLI uses to interface with Dynamo Cloud. You might setup an Ingress, use an `ExternalService` with Istio, or simply port-forward. In our docs, we refer to this externally endpoint with `DYNAMO_SERVER`
```

## Next Steps

After deploying the Dynamo cloud platform, you can:

> 1. Deploy your first inference graph using the [Dynamo Deploy CLI](dynamo_deploy/operator_deployment.md)

Reviewer: ```suggestion
1. Deploy your first inference graph using the [Dynamo CLI](dynamo_deploy/operator_deployment.md)
```

> 2. Monitor your deployments using the Kubernetes dashboard or CLI

Reviewer: ```suggestion
2. Deploy Dynamo LLM pipelines to Kubernetes using the [Dynamo CLI](../../examples/llm/README.md)!
```

> 3. Scale your services as needed using the Dynamo operator

Reviewer: ```suggestion
3. Manage your deployments using the Dynamo CLI
```

> 

Reviewer: ```suggestion

Afterwards, deploy Dynamo LLM pipelines to Kubernetes using this [guide](../../examples/llm/README.md)!
```
